### ISSUE TITLE
Server List Is Not Accessible With Screen Reader on Mac 

### ISSUE THREAD

**Issue by chigkim** [corrados/jamulus#490]
1. Open server list to join.
2. Turn on VoiceOver with command+f5 (or command+triple click power on newer mac with with touch bar)
3. Try navigating the list with tab and arrows
None of the QTreeWidgetItems are not announced as being selected.
However, if you move the mouse cursor to one of the item and press control+option+f5, you'll hear what's under the mouse.
It seems like QTreeWidgetItems are accessible, but they don't appear as children of QTreeWidget to accessibility API.

**Comment by corrados** [comment]
> However, if you move the mouse cursor to one of the item and press control+option+f5, you'll hear what's under the mouse.

I have tried this but it does not work on my Mac. Maybe it is an issue of my old MacOS.

> It seems like QTreeWidgetItems are accessible, but they don't appear as children of QTreeWidget to accessibility API.

That sounds to me like a but in Qt since the QTreeWidgetItems are always children of QTreeWidget, right? Or is there a possibility to add QTreeWidgetItems in a QTreeWidget if they are not children?

**Comment by chigkim** [comment]
Sorry I'm not sure why control+option+f5 didn't work for you.
When a sighted person manually move my mouse and puts it on one of the 
server, and if I press control+option+f5, VO announced the item for me.
So far, my dirty hack around was to add a each item as a button to the 
tree. This has its own problem where you can't expand with right arrow 
to look at the users, but at least people can look at the servers and 
join one.
Inside CConnectDlg::SetServerList
```
for (int i=0; i<lvwServers->columnCount(); i++) {
     lvwServers->setItemWidget(pNewListViewItem, i, new 
QPushButton(pNewListViewItem->text(i)));
}
```
Inside CConnectDlg::SetPingTimeAndNumClientsResult
```
dynamic_cast<QPushButton*>(lvwServers->itemWidget(pCurListViewItem, 
1))->setText(pCurListViewItem->text(1));
dynamic_cast<QPushButton*>(lvwServers->itemWidget(pCurListViewItem, 
2))->setText(pCurListViewItem->text(2));
```

**Comment by corrados** [comment]
I just tried it out with the screen reader which is built-in the Windows 10 operating system. That screen reader can read all list items without any problem. You can even let it read the connected musicians names.

So I assume that we have an issue which is Qt and Mac related. I don't think I can fix anything in the Jamulus software to get it working.

I see that you have implemented a workaround. But unfortunately, this code I cannot use in the official Jamulus version.

**Comment by chigkim** [comment]
Thanks for the info.
Yea I was pretty sure that hack would break UI. I'll look more into it.
For now, I'll incorporate the server list hack to my own repository and 
distribute to screen reader users using Mac.
FYI, we created a Jamulus group for screen reader users on whatsapp 
yesterday, and we already have 25 people, mostly mac users.

**Comment by ann0see** [comment]
@chigkim once we're moving to Qt6 (https://github.com/hoffie/jamulus/tree/Qt6-hoffie-cleanup) could you please test if anything gets better?

**Comment by ann0see** [comment]
You should be able to have a look at the PR‘s by @hoffie tomorrow/at the weekend

**Comment by hoffie** [comment]
> Sure, please ping me when the migration completes.

Hard to tell when the full migration completes, but you can already try builds from PR #2300 here:
[All binaries](https://github.com/jamulussoftware/jamulus/actions/runs/1758667693) (or: [Direct link for Mac](https://github.com/jamulussoftware/jamulus/suites/5089746973/artifacts/151918116)). Only Windows (64bit) and Mac (non-legacy) have changed. If you are using any other platform, then these binaries are unlikely to improve things for you yet.

Note: This is completely untested. Please report back if you see any breakage or -- hopefully -- improvements with regards to screen reader compatibility.

**Comment by chigkim** [comment]
Unfortunately the server list is still broken.
I could be wrong, but I don't think qt6 addressed QTreeWidget and QTreeWidgetItems for Mac.

**Comment by ann0see** [comment]
That’s not good. Is it a known Qt bug or can Jamulus work around it - in a compatible way of course?

**Comment by ann0see** [comment]
If it’s a Qt Bug, maybe an issue can be raised on their side

**Comment by ann0see** [comment]
Hi, Sorry for the maintenance noise here. I’m just into triaging issues. I've just updated the issue description to fit into the template. Hopefully some day this issue will be fixed.

**Comment by ann0see** [comment]
@Michael1972 Potentially we could port the changes by @chigkim once the release is out and publish it as unofficial "side" version.

**Comment by ann0see** [comment]
Maybe https://forum.qt.io/topic/59824/voiceover-interaction-with-wtreewidget-and-qlistwidget-seems-to-be-broken-in-qt-5-5-0 is related..

**Comment by ann0see** [comment]
I think the best way would to switch to another kind of widget.
This would mean that the connect dialog would be redesigned - probably also visually.

**Comment by ann0see** [comment]
What happens if you disable show all musicians? Does it work better then?

**Comment by ann0see** [comment]
> @Michael1972 Potentially we could port the changes by @chigkim once the release is out and publish it as unofficial "side" version.

@Michael1972 would this work for you? 
For Jamulus 4, it seems to be feasible to have a full GUI redesign (https://github.com/orgs/jamulussoftware/discussions/3454). This hopefully makes this issue obsolete. 

I'd highly appreciate if you could then also check out this build for accessibility problems.

**Comment by Michael1972** [comment]
The idea with the inofficiall side-version would be very good. And I definitely will be testing future builds for accessibility. Is it possible to test beta-buillds? I can't build my-self, because I don't have developing-experience. But I would be very glad to test 3.12-beta-build and provide feedback, in case I can download binary in form of app or pkg for mac.

**Comment by ann0see** [comment]
I didn't have time building for now, but I plan to build the QML version for mac in the coming time. 
The other unofficial side build is probably way easier.

**Comment by ann0see** [comment]
Ok. For the upcoming version, we should build an accessible version. The code/diff is already there, so who would be a volunteer to build it for 3.12.0?

**Comment by ann0see** [comment]
@chigkim would you volunteer to update your version once we release 3.12.0?

**Comment by ann0see** [comment]
Otherwise, maybe there's a way to introduce a build time option for the accessible version? I believe it is not easy as it's in the GUI files.

**Comment by pljones** [comment]
Removed from 3.12.0.  Current status appears to be: that there's a patch, https://github.com/jamulussoftware/jamulus/commit/d05319fc455dd5cecb0e8b69e8515d9770f73644, but that caused breakage to the standard UI, so wasn't an acceptable approach.

What's needed here is:
- either a patch that "just works"
- or `Jamulus.pro` option with `#ifdef`-ed code, along with a new Autobuild builder path for the accessible build

Neither of these looks to be available in the immediate term to deliver as part of the release candidate for 3.12.0.

**Comment by ann0see** [comment]
@pljones No. Please see my last PR #3559
